horovod/common/__init__.py

```diff
@@ -46,12 +47,36 @@ def check_extension(ext_name, ext_env_var, pkg_path, *args):
     ctypes.CDLL(os.path.join(os.path.dirname(__file__),
                              'mpi_lib' + get_ext_suffix()), mode=ctypes.RTLD_GLOBAL)
 
-
-def init():
+def init(comm=None):
     """A function that initializes Horovod.
-    """
-    return MPI_COMMON_LIB_CTYPES.horovod_init()
 
+    Args:
+      comm: List specifying ranks for the communicator, relative to the MPI_COMM_WORLD
+        communicator OR the MPI communicator to use. Given communicator will be duplicated.
+        If None, Horovod will use MPI_COMM_WORLD Communicator.
+
+    """
+    if comm is None:
+        comm=[]
+
+    atexit.register(terminate, finalize=True)
```

Comment:
Wouldn't forcing `finalize=True` cause problems for mpi4py, which will try to finalize as well?

Reply:
Idea: should we do `MPI_Finalize()` only if we did `MPI_Init_thread()`, and not provide `finalize` flag altogether, until somebody comes up with use case where it is necessary?

Reply:
Using `finalize=True` doesn't cause problems for mpi4py as it has a check if MPI is already finalized. If I am not mistaken mpi4py does [this](https://github.com/mpi4py/mpi4py/blob/527867aa8415d7086b1d387eb2f76c42ee060efe/src/mpi4py/MPI/atimport.pxi#L205). If you would prefer we only finalize if we called `MPI_Init_thread()` I can do that, it doesn't affect our use case.

Reply:
It's easy to add parameters, but hard to remove. :-)

I think we should finalize only if we initialized, and provide knobs if somebody asks for it.